### DISCUSSION TITLE
first steps to start with custom model

### DISCUSSION THREAD

**Discussion by julen9** [eclipse-glsp/glsp#1042] [Q&A] [answered]
Hi, I want to start customizing the java-emf-theia example, but i can not find how to load my own .ecore model. Please, could you tell me where can I find some documentation or first steps? My first goal is only try to add one more node element to the palette. Thanks a lot in advance!

**Comment by martin-fleck-at** [comment, marked as answer]
Hi @julen9,

I assume you have seen the [README](https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-theia) and the 'Next Steps' they are proposing with the referenced documentation.

In the example repository, we use the [TaskListSourceModelStorage](https://github.com/eclipse-glsp/glsp-examples/blob/9f347e8c2d43d42cf3886e6a531bae3dcf8220a4/project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/model/TaskListSourceModelStorage.java#L22) to properly set up our resource set that is used when loading a model from a URI provided by the `RequestModelAction` from the client. The crucial part here is that the resource set that we use to load the model knows about our custom Ecore-based package so we need to register our package accordingly.

If you want to add one more node element - that does not yet exist in the [model.ecore](https://github.com/eclipse-glsp/glsp-examples/blob/9f347e8c2d43d42cf3886e6a531bae3dcf8220a4/project-templates/java-emf-theia/glsp-server/model/model.ecore) - to the palette, your first step would be to add it to the ecore (or replace it with your own file if you already have one) and then re-generate the model code using the [model.genmodel](https://github.com/eclipse-glsp/glsp-examples/blob/9f347e8c2d43d42cf3886e6a531bae3dcf8220a4/project-templates/java-emf-theia/glsp-server/model/model.genmodel) file. This will create the new sources in the `src-gen` folder from where you can reference them when we translate the semantic model to the GModel. Please make sure that your custom type is somehow part of the "root" element (in the example that is the `TaskList`).

Afterwards, you need to make sure that your semantic element is properly translated to something that GLSP understands, i.e., the [Graphical Model Structure](https://www.eclipse.org/glsp/documentation/gmodel/#graphical-model-structure). One very important aspect is the `type` which is used to map elements to their graphical representation.

1. Specify a custom type for your element to make changes and adaptations easier. Some types are already defined in `TaskListModelTypes`.
2. Adapt the `TaskListToolPaletteItemProvider` to add an entry to the tool palette for your custom node. Here you can do something similar to `createTask` but make sure to use your own type id. That is already sufficient to get the entry into the tool palette. However, GLSP still does not know how to create, map, or render the custom element.
4. Create a custom `AbstractEMFCreateNodeOperationHandler` or your type (the constructor takes the element type id) and register it in the `TaskListDiagramModule`. This command should create the semantic element in the semantic Ecore-based model and the shape in the notation model.
5. Adapt the `TaskListGModelFactory` to ensure that any instances of your class that you find in a `TaskList` root node are properly translated to a GModel representation, e.g., a GNode that has your type.
6. Ensure that on the client side your custom type is properly mapped to a visual representation, e.g., a node with rounded corners.

I've also quickly sketched this on an example branch here: https://github.com/eclipse-glsp/glsp-examples/compare/master...example-javaemftheia-extension 

Hope that gets you started in the right direction.

**Reply by DaveHOnCode** [reply]
Hi @martin-fleck-at,

What do you mean by 

> Please make sure that your custom type is somehow part of the "root" element (in the example that is the `TaskList`)

We are planning to mirgrate from a self programmed webapplication to the java / theia / emf combination. We already replaced the source model with our own custom model, but the resolved semantic element of the notation model does not get filled so far.

Thank you so much!

**Reply by martin-fleck-at** [reply]
What I meant by that was simply that you usually only have one root entry point in Ecore and in your diagram and you should make sure that the other elements are reachable from there by being part of the root tree. In the `TaskList` example both `Task` and `Transition` elements are part of the `TaskList` root element. When we fill the GModel root element with the Ecore root element ([binding](https://github.com/eclipse-glsp/glsp-examples/blob/9f347e8c2d43d42cf3886e6a531bae3dcf8220a4/project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/TaskListDiagramModule.java#L50), [implementation](https://github.com/eclipse-glsp/glsp-examples/blob/9f347e8c2d43d42cf3886e6a531bae3dcf8220a4/project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/model/TaskListGModelFactory.java#L35)) we have a 1:1 mapping between Task -> Node and Transition -> Edge.

> but the resolved semantic element of the notation model does not get filled so far.

I do not fully understand what you mean by that.

**Reply by DaveHOnCode** [reply]
Thank you for your answer @martin-fleck-at !

We replaced parts of the code in the `TaskListGModelFactory.fillRootElement(final EObject semanticModel, final Diagram notationModel, final GModelRoot newRoot)` method with our own code. 

The `notationModel` which is passed into this function does contain a `semanticElement.resolvedSemanticElement` which is `null`.

The `semanticModel` does contain the whole meta model (as expected) althought.

Is there anything else what we have to take care about?

Thanks in advance!

**Reply by martin-fleck-at** [reply]
That is interesting, we set the semantic element on the notation element during [indexing](https://github.com/eclipse-glsp/glsp-server/blob/2bb9e0b545f9f2657b26c8526f16a40ef5d2228e/plugins/org.eclipse.glsp.server.emf/src/org/eclipse/glsp/server/emf/notation/EMFNotationModelIndex.java#L65) in the `EMFNotationModelIndex`. This index implementation is automatically used when your diagram module inherits from `EMFNotationDiagramModule` and you do not re-bind the index yourself.

Does your GModelFactory inherit from EMFNotationGModelFactory? Do you override the `fillRootElement(GmodelRoot)` call or the `fillRootElement(EObject, Diagram, GModelRoot)` call? The [first one](https://github.com/eclipse-glsp/glsp-server/blob/2bb9e0b545f9f2657b26c8526f16a40ef5d2228e/plugins/org.eclipse.glsp.server.emf/src/org/eclipse/glsp/server/emf/notation/EMFNotationGModelFactory.java#L39) has this line `modelState.getIndex().indexAll(notationModel, semanticModel);` that should trigger the correct indexing. Maybe you can debug into that to see why the semantic elements may not be set?

**Reply by DaveHOnCode** [reply]
okay I found the problem. Our model did not implemented the `Identifiable` type. 

Now it seems to work - thank you for the indexing hint, that helped to find the problem

**Comment by esdrascubas** [comment]
Thanks, @martin-fleck-at ,

Very userful answer. I created a new model (ecore and genmodel) and I generated all the classes and implemented a brand new classes to create a new glsp-server with my model. Now, I'm working on the client side. I will be back with new questions.

Thanks a lot!
